### ISSUE TITLE
Cannot destroy resources after moving provider file, even after adding it back

### ISSUE THREAD

**Issue by jackfarzan** [hashicorp/terraform#26734]
### Terraform Version
0.12.29

### Expected Behavior
Resources should be destroyed after moving provider files, or after adding files back. 

### Actual Behavior
```Error: Provider configuration not present``` when attempting to destroy resources. 

### Steps to Reproduce
1. Create some Terraform modules without providers
2. Specify provider when calling the module
3. Move provider into the Terraform module
4. Remove a resource
5. terraform plan
6. Add provider back to original location
7. terraform plan

### Context
I have a central repository of Terraform modules that my team and I manage. These modules are then called in other repositories for separate app projects.

Initially, the central repository did not specify Terraform providers, these providers were specified in the project repos. After realizing this was not a best practice, we moved the providers from the project repos and to the central repo.

However, we did not realize that you need the provider to be able to delete resources. We tried adding the provider file back to the project repos to delete resources, but still we cannot delete the resources through Terraform, we must remove the state manually and delete the resources manually.

I have tried removing and re-importing all resources from the Terraform state, but this hasn't helped.

In short, previous state:

tf-modules-resources
-- dynamodb-table
---- files.tf

resources-project
-- dynamodb_test.tf
-- providers.tf

Current state:
tf-modules-resources
-- dynamodb-table
---- files.tf
---- providers.tf

resources-project
-- dynamodb_test.tf

Trying to even add a dummy `provider aws {}` back to the `resources-project` repo doesn't resolve the issue.

**Comment by jackfarzan** [comment]
I seem to have found a workaround for this issue. I need to add both a dummy `provider aws {}` to the child module, and explicitly provide this dummy in the module provider definition. I can delete things successfully now, but not sure if this is expected behavior, I would think Terraform would be able to inherit the parent module's provider definition.

**Comment by pselle** [comment]
Hi @jackfarzan! It appears you've figured out how to work within your issue. As you figured out, providers are required for Terraform to destroy resources -- providers are the plugin layer that enable Terraform to communicate with services aroud resource creation, updates, _and_ deletion.

In the future, you might find [the community forums](https://www.terraform.io/community.html) helpful for sharing your questions in the future.

As to the question about expected behavior and providers, in Terraform 0.13+ (I see you're using Terraform 0.12 at the moment) we recommend using the `required_providers` block to explicitly require providers for a configuration, which would hopefully help you with this need in the future.

**Comment by jackfarzan** [comment]
Hi @pselle - I am indeed using the `required_providers` block. I am also not sure if my issue is resolved - it was resolved for my single test module, but I am not sure if this will work with my production modules. Is there a way for me to be able to explicitly define the parent provider when instantiating a child module?

**Comment by pselle** [comment]
@jackfarzan I've reopened and marked your issue as "waiting for reproduction" -- my impression was that you'd had an initial question, and then figured out your issue and what remained was a question. 

"Modules without providers" isn't something that happens in Terraform if you have resources -- API calls or any actions that create, update, or destroy resources are done by providers. So if you can provide a simplified configuration that helps demonstrate your issue/question, we can offer more assistance in finding whether or not this is a bug in Terraform, or expected behavior.

**Comment by jackfarzan** [comment]
I have a simplified file structure in my initial issue - I initially did not have a provider definition in my parent module, I only had this definition in my child module. I then decided to add the definition to my parent module, and remove it from the child module. Now, I am unable to destroy instances of the parent module from the child module. This issue occurs both with resources that were created when the child module's provider definition was present, and with resources that were created when there was only a parent module provider, not a child module provider. 

I was under the impression that I only needed prover definitions in my parent module. The documentation I've been reading hasn't been explicit about this. If this is true, I would like to be able to remove the references to the child module provider, and redirect them to the parent module provider.

I can work on creating an example config if it is needed.

**Comment by jackfarzan** [comment]
For example, in my tf-dynamodb-module, I had dynamodb.tf, etc, and no provider definition.

In tf-squad-repo, I had something like:

```
module my_table {
  source = tf-dynamodb-module
...
}

provider aws {
   region = "us-east-1"
   version = "~> 2.7"
}
```

Then, I removed the provider definition there, and added it to the tf-dynamodb-module. Do I still need some empty provider definition?

**Comment by jackfarzan** [comment]
I have not been able to resolve this issue thus far, despite following documented Terraform best practices. Would appreciate input.

**Comment by pselle** [comment]
@jackfarzan If you can provide a simplified configuration and steps for reproduction, it would help me replicate your issue to determine if this is a bug in Terraform (that we could then fix) or if what is happening is expected behavior -- as it is, I can't see how to create a configuration to exactly replicate what you're describing (to ensure we are discussing the same thing).

**Comment by jackfarzan** [comment]
Understood, I will prepare a sample repo as soon as I can.

**Comment by pselle** [comment]
@jackfarzan Thanks so much for making this repo! It's helped clarify your comments and the flow you've been following. 

I was able to hit the error you described! And it is accurate:

```
To work with module.example_module.aws_s3_bucket.bucket its original provider
configuration at module.example_module.provider.aws is required, but it has
been removed.
```

This is an accurate error, as provider configuration is necessary for destruction, and by commenting out/removing the module, we've removed the provider. If the provider is passed to the module by either being available at the parent level, or passed as a proxy provider (link follows), then you will not run into this error.

This is why we don't recommend configuring providers in modules (and in fact, disallow it if the module call has `count` or `for_each` on its declaration), but instead passing providers to the module _if_ it is necessary, per https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly. It is only necessary if there is not a provider at the parent level that can fulfill the child module's requests.

I'm going to close this issue, because as far as I can tell the error message are behaving as expected, and I haven't been able to find a bug in Terraform related to this investigation.

**Comment by jackfarzan** [comment]
@pselle can you clarify exactly what you mean by the provider being available at the parent level? I thought this meant including the provider in the module itself.
